### ISSUE TITLE
LockError: Unable to fully release the lock on resource

### ISSUE THREAD

**Issue by nmargaritis** [mike-marcacci/node-redlock#65]
After upgrading to v4, I started getting this error `LockError: Unable to fully release the lock on resource`.

`Redis server v=5.0.7`. (appears to have the same problems in newer versions as well)

I rolled back to version 3.1.2 for now.

This could be related with [#39](https://github.com/mike-marcacci/node-redlock/issues/39)

**Comment by YSMull** [comment]
me too!

**Comment by maxs15** [comment]
Same issue on my side!

**Comment by mike-marcacci** [comment]
Hi all! Just now getting a chance to look at this. I've combed through the logic and tests and I haven't found anything suspicious.

A few questions:
1. Are these running against redis instances or clusters?
2. Are there multiple instances/clusters?
3. Do you have any non-default settings on your clusters that effect behavior?
4. What client lib are you using (ioredis/node-redis)?
5. Do you have any non-default settings on your client lib?

Is there any chance one of you can create a minimal reproduction of this, perhaps using docker-compose to set up the environment?

**Comment by nmargaritis** [comment]
> Hi all! Just now getting a chance to look at this. I've combed through the logic and tests and I haven't found anything suspicious.
> 
> A few questions:
> 
> 1. Are these running against redis instances or clusters?
> 2. Are there multiple instances/clusters?
> 3. Do you have any non-default settings on your clusters that effect behavior?
> 4. What client lib are you using (ioredis/node-redis)?
> 5. Do you have any non-default settings on your client lib?
> 
> Is there any chance one of you can create a minimal reproduction of this, perhaps using docker-compose to set up the environment?

1. Instances
2. Single instance in my usecase
3. Nothing special
4. Ioredis `"ioredis": "4.15.1"`
5. I am providing my redlock + ioredis init
```
        this.redlock = new Redlock([this.store], {
            driftFactor: 0.01, // time in ms
            retryCount: 25,
            retryDelay: 300, // time in ms
            retryJitter: 200 // time in ms
        });

this.store = new Redis({
            host: process.env.REDIS_HOST,
            port: process.env.REDIS_PORT,
            password: process.env.REDIS_PASSWORD,
            connectionName: 'translations',
            keepAlive: true,
            retryStrategy: function(times) {
                var delay = Math.min(times * 3000, 60000);
                return delay;
            },
            reconnectOnError: err => {
                var targetError = 'READONLY';
                if (err.message.slice(0, targetError.length) === targetError) {
                    // Only reconnect when the error starts with "READONLY"
                    return true; // or `return 1;`
                }
            }
        });
```

Not sure about the latter. Btw Redis is on AWS (cache), if this is helpful.

**Comment by amirzandi** [comment]
I still have problem with following error: 
`{ LockError: Unable to fully release the lock on resource "lock:tournament:winner:289247".
0|server  |     at Command.loop [as callback] (/var/www/tornoq/node_modules/redlock/redlock.js:264:18)
0|server  |     at normal_reply (/var/www/tornoq/node_modules/redis/index.js:726:21)
0|server  |     at RedisClient.return_reply (/var/www/tornoq/node_modules/redis/index.js:824:9)
0|server  |     at JavascriptRedisParser.returnReply (/var/www/tornoq/node_modules/redis/index.js:192:18)
0|server  |     at JavascriptRedisParser.execute (/var/www/tornoq/node_modules/redis-parser/lib/parser.js:574:12)
0|server  |     at Socket.<anonymous> (/var/www/tornoq/node_modules/redis/index.js:274:27)
0|server  |     at Socket.emit (events.js:193:13)
0|server  |     at addChunk (_stream_readable.js:296:12)
0|server  |     at readableAddChunk (_stream_readable.js:277:11)
0|server  |     at Socket.Readable.push (_stream_readable.js:232:10)
0|server  |     at TCP.onStreamRead (internal/stream_base_commons.js:150:17)
0|server  |   name: 'LockError',
0|server  |   message:
0|server  |    'Unable to fully release the lock on resource "lock:tournament:winner:289247".',
0|server  |   attempts: undefined }`


Please help

**Comment by amirzandi** [comment]
@mike-marcacci please check this error.

**Comment by amirzandi** [comment]
@mike-marcacci Why dont you follow up issues? we are facing much problems with this error.... If you are not able to resolve it, please let us know to change the component.

**Comment by nmargaritis** [comment]
> @mike-marcacci Why dont you follow up issues? we are facing much problems with this error.... If you are not able to resolve it, please let us know to change the component.

@amirzandi Hi, have you tried to roll back to version `3.1.2` until this issue is resolved? 
Maybe the maintainer does not have enough information to debug this issue, for now.

**Comment by amirzandi** [comment]
I finally found out the problem and I resolved it myself. I will have some plan to use another component in order to get rid of such issues.

**Comment by nmargaritis** [comment]
> I finally found out the problem and I resolved it myself. I will have some plan to use another component in order to get rid of such issues.

@amirzandi  Could you please create a pull request so that the fix is merged back to the project, and anyone can use it.

**Comment by amirzandi** [comment]
I only increased the lock time to 60000 milliseconds and it is now working fine.

**Comment by seanfisher** [comment]
I had this same error in version 4.1.0, which went away after I removed the `keyPrefix` option from my `Ioredis` (v4.16.2) configuration.

**Comment by b00tsy** [comment]
For me this goes away if going back to 3.1.2. But having a look at the timing it appears that this "new" error message pops up when the lock expired. Eg. if locking for 250 ms and the actual processing while the lock is acquired takes 300 ms an error gets thrown. Maybe this was enforced with 4.0.0 and before was handled silently?

**Comment by mike-marcacci** [comment]
Hey folks, sorry for the slow replies: I've been unbelievably busy this year so far. I think that @b00tsy may be onto something here, especially given @amirzandi's workaround.

Version 3 _should_ have rejected with the same error, but perhaps it wasn't? Unlocking was _originally_ a best-effort operation that wouldn't report an error, but that was changed quite early on.

For those of you who have reverted to version 3, can you watch your redis instances for key expirations? Something like this:

```
config set notify-keyspace-events Ex
psubscribe __keyevent@0__:expired
```

If all is well and you are acquiring or extending locks with adequate durations, you shouldn't see any keys that match your locks expiring.

**Comment by b00tsy** [comment]
Well I print the time my tasks have taken and sometimes they’re above the ttl I have set for the lock, so I know quite sure, that the error message comes when the key in redis probably has expired. In fact the error gave me a clue that something in my processing pipeline wasn’t running smooth of which I was not aware of before.

The question probably is whether unlock should result in an error if the key expired or whether it should maybe return 0 or 1 (or true or false) depending on whether unlocking was successful and if not print a warning or info message that the key expired before it could be unlocked...

**Comment by bohendo** [comment]
I can reproduce this issue w 100% reliability w these steps:
 1. process 1 requests lock & then never unlocks
 2. process 2 requests lock then it's acquired once process 1's expires
 3. process 2 unlocks after <= the timeout & this error gets thrown.
Logs from our locking service (using ioredis@4.16.3 & redlock@4.1.0) that demonstrates this:
```
2020-05-21T09:15:05.159Z [LockService] Acquiring lock for 0x055b7874f51964fD587969D6666474e99Ac4317F (TTL: 10000 ms)
2020-05-21T09:15:05.159Z [LockService] Acquired lock for 0x055b7874f51964fD587969D6666474e99Ac4317F after waiting 0 ms
2020-05-21T09:15:05.217Z [LockService] Waiting on lock for 0x055b7874f51964fD587969D6666474e99Ac4317F (locked: 0x055b..317F)
2020-05-21T09:15:15.176Z [LockService] Acquired lock for 0x055b7874f51964fD587969D6666474e99Ac4317F after waiting 9959 ms
2020-05-21T09:15:18.665Z [LockService] Releasing lock for 0x055b7874f51964fD587969D6666474e99Ac4317F after 3489 ms
2020-05-21T09:15:18.669Z [LockService] Error while unlocking 0x055b7874f51964fD587969D6666474e99Ac4317F: Unable to fully release the lock on resource "0x055b7874f51964fD587969D6666474e99Ac4317F".
```

**Comment by mike-marcacci** [comment]
@bohendo do you only experience this with these are _separate_ processes? Following your algorithm does not produce any issues in my tests:

```js
const Redlock = require('./redlock');
const IORedis = require('ioredis');

const client = new IORedis();
const redlock = new Redlock([client], {
  retryCount: 10,
  retryDelay: 100,
  retryJitter: 50
});


async function main (){
  console.log("a");
  await redlock.lock("regression", 200);
  console.log("b");
  const second = await redlock.lock("regression", 400);
  console.log("c");
  await second.unlock();
  console.log("d");
}

main().then(console.log, console.error);
```

And just to be 100% sure, there appear to be ~3.5 seconds in your logs between acquiring the lock and releasing it: I assume the lock duration is greater than this?

**Comment by stavalfi** [comment]
same here. the error occurs sometimes but until now I only lock in the same process. any solution? alternatives?

**Comment by nmargaritis** [comment]
The only input I can provide is that, you should not simply increase the lock time/timeout blindly. This is not really a solution and it can only mask other problems of the library or your system. If there is something wrong with your system, this can result in lock time exceptions. cause if your lengthy processes take more time than the total lock acquiring time, then any other attempt to acquire the lock will expire and throw the error. **Such cases should be considered as false positives** as regards to this issue I initially opened. As they are related with misconfiguration or bugs in the system using the lock mechanism.

For my use-case, the system was tested; with each operation reaching a lock total time of 1-2 seconds as a maximum. After upgrading we started hitting the 30seconds auto-expiry timeout, consistently. The only workaround so far was to roll back the redlock version.

I believe that there is something wrong with the changed introduced
[here](https://github.com/mike-marcacci/node-redlock/blob/master/redlock.js#L255)

**Comment by gnitsenko93** [comment]
@mike-marcacci Here is a script for the bug reproduction:

```
'use strict';

const Redis = require('ioredis');
const Redlock = require('redlock');

const TIMEOUT_SEC = 2000;
const ADDITIONAL_TIMEOUT = 100;
const KEY = 'key';

const redis = new Redis();
const redlock = new Redlock([redis]);

redlock.lock(KEY, TIMEOUT_SEC)
    .then(lock => {
        console.log(`resource is locked, key=[${KEY}]`);

        setTimeout(() => {
            lock.unlock()
                .then(() => {
                    console.log(`resource is unlocked, key=[${KEY}]`);
                    process.exit(0);
                })
                .catch(error => {
                    console.log(`error on resource unlocking, error=[${error}]`);
                    process.exit(1);
                })
            
        }, TIMEOUT_SEC + ADDITIONAL_TIMEOUT);
    });
```

You can try run it with different library tags and make sure it works for 3.1.2 but not for 4.1.0.

I personally used Redis server version 3.2.12 into a docker container and ioredis 4.14.2.

My suggestion here is that you have broken that functionality by changing Lua scripts for locking and unlocking.

Seems like a critical bug since an application code throws an error on an attempt to unlock a resource after a timeout has exceeded.

**Comment by gnitsenko93** [comment]
@mike-marcacci So the problem is that in 3.1.2:
1. Lock script sets a value with an expiration timeout
2. Unlock script deleted a value if it exists and returned 0 otherwise
3. There was a check:
```
if(response === 0 || response === 1)	
				votes++;
```
that passed for a case if an unlocked value does not exist and incremented `votes` variable, so that the error was not thrown.

As for 4.1.0 the algorithm has been broken:
...
3. There is the changed condition:
```
if(response === resource.length || response === '' + resource.length)
				votes++;
```
that won't passed for the case.

**Comment by mike-marcacci** [comment]
@gnitsenko93 perhaps I'm missing something here (I'm still on my first coffee), but in your reproduction the lock has already expired and the keys have been evicted by redis... so calling unlock _should_ error. This way your application can alert you to the fact that the lock expired and became re-acquirable _before_ you unlocked it.

If this is the scenario that is being described in this issue, this is expected behavior in v4 as described in [CHANGELOG.md](https://github.com/mike-marcacci/node-redlock/blob/master/CHANGELOG.md#v400). This would make sense given my inability to reproduce an issue that folks are clearly seeing in the wild.

Now, perhaps this isn't a scenario that you would consider an error (and initially I didn't either, given this library's earlier behavior). However, given that the purpose of this library is to guarantee exclusivity of a particular resource, I would consider it an error where the specified expiration breaks this guarantee.

**Comment by gnitsenko93** [comment]
@mike-marcacci Got ya. Closing my pull request.

**Comment by gnitsenko93** [comment]
@mike-marcacci Still there is a small inaccuracy in the changelog.

> Change behavior of `unlock` to return an error if **one** or more entries cannot be removed.

Should be actually:

> Change behavior of `unlock` to return an error if **zero** or more entries cannot be removed.

**Comment by SimenB** [comment]
> Now, perhaps this isn't a scenario that you would consider an error (and initially I didn't either, given this library's earlier behavior). However, given that the purpose of this library is to guarantee exclusivity of a particular resource, I would consider it an error where the specified expiration breaks this guarantee.

@mike-marcacci could the error returned get more details? Right now it says `Unable to fully release the lock on resource` but if it's due to the resource being gone (or something else) it'd be great if the error said so instead of a more generic "failed"

**Comment by ostar0816** [comment]
I think we could just catch that error and handle it not to propagate because it doesn't harm anything just like a warning. We just need to set `ttl` properly according to our needs and it makes sense that the lock is expired if ttl is passed.

**Comment by hisabimbola** [comment]
I started having this issue too. We have been on version 4 for a while but today out of no where, I started to get this error where unable to fully release the lock. I listened for expired event and do not see any event with the locking key.

**Comment by winnochan** [comment]
I now also see the bug in 4.2.0, I want to know which version should I use to avoid this bug?

**Comment by gregbarcza** [comment]
+1

**Comment by Bhautik0110** [comment]
Same issue when upgrading to the latest.

**Comment by Dakuan** [comment]
ditto

**Comment by youkaisteve** [comment]
I have the same issue when acquire with a less ttl than task executing time.

**Comment by mike-marcacci** [comment]
Hi folks! I'm going to go ahead and close this, as the newest published versions are effectively full rewrites and any residual bugs from the former versions are unlikely to persist. It's still not entirely clear whether there is a but here at all or just expected behavior when a routine takes longer than the lock. Regardless, if anybody encounters this again, please open a new issue that we can constrain to a particular version.

**Comment by nexun** [comment]
> Este error ocurre cuando intentamos liberar un bloqueo que ya caducó.
> 
> Una condición como la siguiente al liberar el bloqueo ayudaría
> 
> ```
> if (lock.expiration < Date.now() && lock.expiration != 0) {
>   lock.unlock()
> }
> ```

shouldn't it be lock.expiration **>** Date.now()?

**Comment by and-kost** [comment]
> > ```
> > if (lock.expiration < Date.now() && lock.expiration != 0) {
> >   lock.unlock()
> > }
> > ```
> 
> shouldn't it be lock.expiration **>** Date.now()?

I suppose It should be >

**Comment by dasbts** [comment]
> Hi folks! I'm going to go ahead and close this, as the newest published versions are effectively full rewrites and any residual bugs from the former versions are unlikely to persist. It's still not entirely clear whether there is a but here at all or just expected behavior when a routine takes longer than the lock. Regardless, if anybody encounters this again, please open a new issue that we can constrain to a particular version.

Not true... :(
